File: frontend/src/api/compatibilityRules.ts
/**
 * 兼容性规则引擎 API（/api/compatibility-rules）。
 * 声明式 WHEN(条件)→THEN(动作) 规则，跨 KP 配件库 / 料号库 / 基准机箱 / 商机维度求值。
 * type: require(必配) / exclude(互斥) / derive(派生) / filter(过滤) / recommend(推荐)
 * body: { when:{all/any:[{field,op,value}]}, then:{action,...}, desc }
 */
import axios from 'axios'

const RESP = <T>(p: Promise<{ data: T }>) => p.then(r => r.data)

export type RuleType = 'require' | 'exclude' | 'derive' | 'filter' | 'recommend'
export type RuleStatus = 'draft' | 'testing' | 'active' | 'archived'

export interface CompatibilityRule {
  id: number
  domain: string
  type: RuleType
  name: string
  scope: Record<string, any> | null
  body: Record<string, any> | null       // { when, then, desc }
  status: RuleStatus
  version: number
  hit_count: number
  last_hit_at: string | null
  description?: string | null
  change_reason?: string | null
}

export const compatibilityRulesApi = {
  list: (params?: { type?: RuleType; status?: RuleStatus }) =>
    RESP<{ rules: CompatibilityRule[] }>(axios.get('/api/compatibility-rules/', { params })),
  get: (id: number) => RESP<CompatibilityRule>(axios.get(`/api/compatibility-rules/${id}`)),
  create: (data: Partial<CompatibilityRule> & { type: RuleType; name: string }) =>
    RESP<CompatibilityRule>(axios.post('/api/compatibility-rules/', data)),
  update: (id: number, data: Partial<CompatibilityRule>) =>
    RESP<CompatibilityRule>(axios.put(`/api/compatibility-rules/${id}`, data)),
  setStatus: (id: number, status: RuleStatus) =>
    RESP<CompatibilityRule>(axios.post(`/api/compatibility-rules/${id}/status`, { status })),
  remove: (id: number) =>
    RESP<{ success: boolean }>(axios.delete(`/api/compatibility-rules/${id}`)),
  reset: () =>
    RESP<{ reset: boolean; count: number }>(axios.post('/api/compatibility-rules/reset')),
  recordHit: (id: number) =>
    RESP<{ id: number; hit_count: number; last_hit_at: string }>(axios.post(`/api/compatibility-rules/${id}/hit`)),
  stats: (id: number) =>
    RESP<{ hit_count: number; last_hit_at: string | null }>(axios.get(`/api/compatibility-rules/${id}/stats`)),
}

File: backend/app/api/compatibility_rules.py
"""API endpoints for compatibility rules (兼容性规则引擎).

type: require(必配) / exclude(互斥) / derive(派生) / filter(过滤) / recommend(推荐)
status: draft / testing / active / archived（只有 active 被执行器引用）
body: {when:{all/any:[{field,op,value}]}, then:{action,...}, desc}
"""
from fastapi import APIRouter, HTTPException
from app.repository.compatibility_rule_repo import CompatibilityRuleRepository

router = APIRouter(prefix="/api/compatibility-rules", tags=["compatibility-rules"])

_VALID_STATUS = {"draft", "testing", "active", "archived"}
_VALID_TYPE = {"require", "exclude", "derive", "filter", "recommend"}


@router.get("/")
def list_rules(type: str = None, status: str = None, category: str = None):
    repo = CompatibilityRuleRepository()
    try:
        return {"rules": repo.list(type=type, status=status, category=category)}
    finally:
        repo.close()


@router.post("/reset")
def reset_rules():
    """清空并重置为默认规则（规则版本迭代后让用户一键更新到最新 seed）。返回插入条数。"""
    repo = CompatibilityRuleRepository()
    try:
        count = repo.reset_to_defaults()
        return {"reset": True, "count": count}
    finally:
        repo.close()


@router.get("/{rule_id}")
def get_rule(rule_id: int):
    repo = CompatibilityRuleRepository()
    try:
        r = repo.get(rule_id)
        if not r:
            raise HTTPException(404, f"Rule {rule_id} not found")
        return r
    finally:
        repo.close()


@router.post("/")
def create_rule(data: dict):
    for k in ("type", "name"):
        if not data.get(k):
            raise HTTPException(400, f"Missing field: {k}")
    if data["type"] not in _VALID_TYPE:
        raise HTTPException(400, f"Invalid type, must be one of {sorted(_VALID_TYPE)}")
    repo = CompatibilityRuleRepository()
    try:
        return repo.create(data, operator=data.get("operator", "system"))
    finally:
        repo.close()


@router.put("/{rule_id}")
def update_rule(rule_id: int, data: dict):
    repo = CompatibilityRuleRepository()
    try:
        r = repo.update(rule_id, data, operator=data.get("operator", "system"))
        if not r:
            raise HTTPException(404, f"Rule {rule_id} not found")
        return r
    finally:
        repo.close()


@router.post("/{rule_id}/status")
def set_status(rule_id: int, data: dict):
    status = data.get("status")
    if status not in _VALID_STATUS:
        raise HTTPException(400, f"Invalid status, must be one of {sorted(_VALID_STATUS)}")
    repo = CompatibilityRuleRepository()
    try:
        r = repo.set_status(rule_id, status, operator=data.get("operator", "system"))
        if not r:
            raise HTTPException(404, f"Rule {rule_id} not found")
        return r
    finally:
        repo.close()


@router.delete("/{rule_id}")
def delete_rule(rule_id: int):
    repo = CompatibilityRuleRepository()
    try:
        if not repo.delete(rule_id):
            raise HTTPException(404, f"Rule {rule_id} not found")
        return {"success": True}
    finally:
        repo.close()


@router.post("/{rule_id}/hit")
def record_hit(rule_id: int):
    """记录规则命中（越跑越聪明）。执行器每次命中调用。"""
    repo = CompatibilityRuleRepository()
    try:
        r = repo.record_hit(rule_id)
        if not r:
            raise HTTPException(404, f"Rule {rule_id} not found")
        return r
    finally:
        repo.close()


@router.get("/{rule_id}/stats")
def rule_stats(rule_id: int):
    repo = CompatibilityRuleRepository()
    try:
        return repo.stats(rule_id)
    finally:
        repo.close()
